File: BowlingCompanion/Sources/SharedModelsLibrary/Alley.swift
import Foundation
import StringsLibrary

public struct Alley: Sendable, Identifiable, Hashable, Codable {
	public let id: UUID
	public let name: String
	public let address: String?
	public let material: Material
	public let pinFall: PinFall
	public let mechanism: Mechanism
	public let pinBase: PinBase

	public init(
		id: UUID,
		name: String,
		address: String?,
		material: Material,
		pinFall: PinFall,
		mechanism: Mechanism,
		pinBase: PinBase
	) {
		self.id = id
		self.name = name
		self.address = address
		self.material = material
		self.pinFall = pinFall
		self.mechanism = mechanism
		self.pinBase = pinBase
	}
}

extension Alley {
	public enum Material: Int, Sendable, Identifiable, CaseIterable, Codable, CustomStringConvertible {
		case synthetic = 0
		case wood = 1
		case unknown = -1

		public var id: Int { rawValue }
		public var description: String {
			switch self {
			case .synthetic: return Strings.Alley.Properties.Material.synthetic
			case .wood: return Strings.Alley.Properties.Material.wood
			case .unknown: return Strings.unknown
			}
		}
	}
}

extension Alley {
	public enum PinFall: Int, Sendable, Identifiable, CaseIterable, Codable, CustomStringConvertible {
		case freefall = 0
		case strings = 1
		case unknown = -1

		public var id: Int { rawValue }
		public var description: String {
			switch self {
			case .freefall: return Strings.Alley.Properties.PinFall.freefall
			case .strings: return Strings.Alley.Properties.PinFall.strings
			case .unknown: return Strings.unknown
			}
		}
	}
}

extension Alley {
	public enum Mechanism: Int, Sendable, Identifiable, CaseIterable, Codable, CustomStringConvertible {
		case dedicated = 0
		case interchangeable = 1
		case unknown = -1

		public var id: Int { rawValue }
		public var description: String {
			switch self {
			case .dedicated: return Strings.Alley.Properties.Mechanism.dedicated
			case .interchangeable: return Strings.Alley.Properties.Mechanism.interchangeable
			case .unknown: return Strings.unknown
			}
		}
	}
}

extension Alley {
	public enum PinBase: Int, Sendable, Identifiable, CaseIterable, Codable, CustomStringConvertible {
		case black = 0
		case white = 1
		case other = 2
		case unknown = -1

		public var id: Int { rawValue }
		public var description: String {
			switch self {
			case .black: return Strings.Alley.Properties.PinBase.black
			case .white: return Strings.Alley.Properties.PinBase.white
			case .other: return Strings.other
			case .unknown: return Strings.unknown
			}
		}
	}
}

File: android/core/model/src/main/java/ca/josephroque/bowlingcompanion/core/model/Alley.kt
package ca.josephroque.bowlingcompanion.core.model

import android.os.Parcelable
import ca.josephroque.bowlingcompanion.core.model.utils.SortableByUUID
import java.util.UUID
import kotlinx.parcelize.Parcelize

@JvmInline
@Parcelize
value class AlleyID(val value: UUID) : Parcelable {
	override fun toString(): String = value.toString()
	companion object {
		fun randomID(): AlleyID = AlleyID(UUID.randomUUID())
		fun fromString(string: String): AlleyID = AlleyID(UUID.fromString(string))
	}
}

data class AlleyListItem(
	val alleyId: AlleyID,
	val name: String,
	val material: AlleyMaterial?,
	val pinFall: AlleyPinFall?,
	val mechanism: AlleyMechanism?,
	val pinBase: AlleyPinBase?,
) : SortableByUUID {
	override val id: UUID
		get() = alleyId.value
}

data class AlleyDetails(
	val id: AlleyID,
	val name: String,
	val material: AlleyMaterial?,
	val pinFall: AlleyPinFall?,
	val mechanism: AlleyMechanism?,
	val pinBase: AlleyPinBase?,
)

data class AlleyCreate(
	val id: AlleyID,
	val name: String,
	val material: AlleyMaterial?,
	val pinFall: AlleyPinFall?,
	val mechanism: AlleyMechanism?,
	val pinBase: AlleyPinBase?,
	val lanes: List<LaneListItem>,
)

data class AlleyUpdate(
	val id: AlleyID,
	val name: String,
	val material: AlleyMaterial?,
	val pinFall: AlleyPinFall?,
	val mechanism: AlleyMechanism?,
	val pinBase: AlleyPinBase?,
	val lanes: List<LaneListItem>,
) {
	data class Properties(
		val id: AlleyID,
		val name: String,
		val material: AlleyMaterial?,
		val pinFall: AlleyPinFall?,
		val mechanism: AlleyMechanism?,
		val pinBase: AlleyPinBase?,
	)
}

enum class AlleyMaterial {
	SYNTHETIC,
	WOOD,
}

enum class AlleyMechanism {
	DEDICATED,
	INTERCHANGEABLE,
}

enum class AlleyPinBase {
	BLACK,
	WHITE,
	OTHER,
}

enum class AlleyPinFall {
	FREE_FALL,
	STRINGS,
}
